astropy/cosmology/io/core.py

```diff
@@ -0,0 +1,291 @@
+
+    # select arguments from mapping that are in the cosmo's signature.
+    ba = cosmology._init_signature.bind_partial()  # blank set of args
+    ba.apply_defaults()  # fill in the defaults
+    for k in cosmology._init_signature.parameters.keys():  # iter thru sig
+        if k in params:  # transfer argument, if in params
+            ba.arguments[k] = params.pop(k)
+
+    # deal with remaining params. If there is a **kwargs use that, else
+    # allow to transfer to metadata. Raise TypeError if can't.
+    lastp = tuple(cosmology._init_signature.parameters.values())[-1]
+    if lastp.kind == 4:  # variable keyword-only
+        ba.arguments[lastp.name] = params
+    elif move_to_meta:  # prefers current meta, which was explicitly set
+        meta = ba.arguments["meta"] or {}  # (None -> dict)
+        ba.arguments["meta"] = {**params, **meta}
+    elif bool(params):
+        raise TypeError(f"There are unused parameters {params}.")
+    # else: pass  # no kwargs, no move-to-meta, and all the params are used
+
+    return cosmology(*ba.args, **ba.kwargs)
+
+
+def to_mapping(cosmology):
+    """Return the cosmology class, inputs, and metadata as a dict.
+
+    Parameters
+    ----------
+    cosmology : `~astropy.cosmology.Cosmology` subclass instance
+
+    Returns
+    -------
+    dict
+        with key-values for the cosmology parameters and also:
+
+        - 'cosmology' : the class
+        - 'meta' : the contents of the cosmology's metadata attribute
+
+    Examples
+    --------
+    A Cosmology as a mapping will have the cosmology's name and
+    parameters as items, and the metadata as nested dictionary.
+
+        >>> from astropy.cosmology import io, Planck18
+        >>> io.to_mapping(Planck18)
+        {'cosmology': <class 'astropy.cosmology.core.FlatLambdaCDM'>,
+         'name': 'Planck18', 'H0': 67.66, 'Om0': 0.30966, 'Tcmb0': 2.7255,
+         'Neff': 3.046, 'm_nu': [0.0, 0.0, 0.06], 'Ob0': 0.04897,
+         'meta': ...
+
+    See Also
+    --------
+    astropy.cosmology.io.from_mapping : Load Cosmology from dictionary.
+    astropy.cosmology.io.from_table : Load Cosmology from |QTable|.
+    astropy.cosmology.io.to_table : Represent as |QTable|.
+    astropy.cosmology.Cosmology.write : Has a ``to_mapping`` convenience method.
+    """
+    m = {}
+    # start with the cosmology class & name
+    m["cosmology"] = cosmology.__class__
+    m["name"] = cosmology.name  # here only for dict ordering
+    # get all the immutable inputs
+    m.update({k: v for k, v in cosmology._init_arguments.items()
+              if k not in ("meta", "name")})
+    # add the mutable metadata
+    m["meta"] = copy.deepcopy(cosmology.meta)
+
+    return m
+
+
+def from_table(table, index=None, *, move_to_meta=False):
+    """Instantiate a `~astropy.cosmology.Cosmology` from a |QTable|.
+
+    Parameters
+    ----------
+    cosmology : `~astropy.cosmology.Cosmology` class
+    table : `~astropy.table.QTable`
+    index : int or None, optional
```

Comment:
I considered that. Ultimately, I found including ``index`` as a kwarg more convenient for both the user and contributor. For the user, it's of similar convenience, but more readable.
For contributors, or perhaps just a user writing a custom read/write method, it's nice to have the index logic in multi-row table baked in. It definitely saves on re-writing.
While currently only ECSV uses this method, I want to eventually have (possibly a good PR for first-time contributors) read/write for latex and HTML. Other formats I'm interested in are: CLASS's and CAMB's ``.ini`` formats, ``Emcee``'s and ``MontePython``'s various save formats. The latter two are multi-row (because MCMC) formats.

I do think that your suggestion of ``table['name']==name`` is a good one for avoiding creating an indexed copy of the table. Instead I can just do ``np.where(table['name']==name)`` (with an uniqueness check). I'll workshop that, thanks!

Reply:
> Can also be str. 

Done!